simba/optimization.py

```diff
@@ -122,3 +125,269 @@ def service_optimization(schedule, scenario, args):
+        add_depot_trip(last_trip.departure_name, last_trip.arrival_name, last_trip, depot_trips)
+
+        trip_list = []
+        # look at all trips except first and last
+        for trip in rot.trips[1:-1]:
+            trip_list.append(trip)
+            # check if intermediate trip starts or ends in depot
+            departure_station = schedule.stations.get(trip.departure_name)
+            arrival_station = schedule.stations.get(trip.arrival_name)
+            depot_station = None
+            trip_station = None
+            if departure_station is not None and departure_station["type"] == "deps":
+                depot_station = trip.departure_name
+                trip_station = trip.arrival_name
+            elif arrival_station is not None and arrival_station["type"] == "deps":
+                depot_station = trip.arrival_name
+                trip_station = trip.departure_name
+            if depot_station is not None:
+                # trip starts or ends in depot: save trip in depot_trips
+                add_depot_trip(trip_station, depot_station, trip, depot_trips)
+                # dummy trip from depot to same depot for lookup
+                depot_trip = Trip(
+                    rotation=rot,
+                    departure_time=trip.arrival_time,
+                    departure_name=depot_station,
+                    arrival_time=trip.arrival_time,
+                    arrival_name=depot_station,
+                    distance=0.1,  # must not be zero?
+                    line=None,
+                    temperature=trip.temperature,
+                    height_difference=0,
+                    level_of_loading=0,
+                    mean_speed=18,
+                )
+                add_depot_trip(depot_station, depot_station, depot_trip, depot_trips)
+        # all trips done
+
+        # make note of trips from negative depot rotations
+        qualified_id = negative_rotations is None or rot_id in negative_rotations
+        if rot.charging_type == "depb" and qualified_id:
+            trips[rot_id] = trip_list
+    return (trips, depot_trips)
+
+
+def add_depot_trip(station_name, depot_name, trip, depot_trips):
+    if station_name in depot_trips:
+        if depot_name in depot_trips[station_name]:
+            # trip from station to depot already known: use trip with lesser distance
+            if trip.distance < depot_trips[station_name][depot_name].distance:
+                depot_trips[station_name][depot_name] = trip
+        else:
+            # trip from station to depot unknown: save trip as reference
+            depot_trips[station_name][depot_name] = trip
+    else:
+        # no trip to any depot known for this station: save trip as reference
+        depot_trips[station_name] = {depot_name: trip}
+    return depot_trips
+
+
+def generate_depot_trip(station, depot_trips, default_depot_distance, default_mean_speed):
+    """
```

Comment:
generates a depot_trip_data_dict or something along those lines but not a trip which is a defined SimBA class